fix: ggml: fix vulkan-shaders-gen build

sparkleholic opened the pull request:
The vulkan-shaders-gen target was not being built correctly in case of cross-compilation.
Other outputs need to be built for the cross compile target, but vulkan-shaders-gen needs to be built for the host.



- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

sparkleholic committed:
fix: ggml: fix vulkan-shaders-gen build

The vulkan-shaders-gen target was not being built correctly
in case of cross-compilation.
Other outputs need to be built for the cross compile target,
but vulkan-shaders-gen needs to be built for the host.

sparkleholic commented:
@jeffbolznv , @0cc4m 
This fix is ​​needed for cross-compilation.
`vulkan-shaders-gen` is needed at build time, so `arch` should be the same as `host pc`, but other binaries (libggml, libllama, etc.) are needed at runtime.
Therefore, CMAKE compilation variables that apply when building the llama.cpp project should not apply when building vulkan-shaders-gen.

0cc4m commented:
@sparkleholic @jeffbolznv I understand, I'm sorry that a few PRs are getting delayed because of me. I'm currently ill, but getting better and I should be able to catch up with all of them later this week.

Maybe you can resolve the merge conflict already?

jeffbolznv commented:
No rush, get well!

I'm not an expert at cmake or cross compilation, so I don't think I should review the change.

sparkleholic commented:
@0cc4m 
master rebased to resolve conflicts.
I didn't mean to rush you. I sincerely hope you feel better soon. :)

0cc4m commented:
I can confirm this still works as usual on my system, but I'm also not a CMake expert. Maybe @bandoti can take a look?

bandoti commented:
I am happy to take a look. The one thing that comes to mind here after initial look through is being able to customize the compiler in this case. In particular, `gcc` and `g++` are hardcoded.

It would be worthwhile to add a default toolchain file for the build and allow one to customize by passing `-DGGML_SHADERS_GEN_CROSS_TOOLCHAIN=...` (or similar)

This would allow one to use a desire toolchain—maybe clang, etc.

sparkleholic committed:
refactor: ggml: Improve vulkan-shaders-gen toolchain setup

- Add GGML_SHADERS_GEN_TOOLCHAIN CMake option.
- Auto-detect host toolchain if not set.

sparkleholic commented:
@bandoti 
Thank you for the valuable feedback.

I have updated the code to handle the `GGML_SHADERS_GEN_TOOLCHAIN` variable. When this variable is provided, the specified file is used. If it is not set, the code generates a temporary host toolchain file by detecting the host environment's compiler.

Please review the updated code once again.

0cc4m commented:
@bandoti Can you take another look?

bandoti commented:
@sparkleholic Sorry for the delay! I had these review comments in but forgot to submit them. 🤦‍♂️

bandoti commented:
@sparkleholic Just following up on this change. If you need further assistance please let me know I am happy to help.

sparkleholic commented:
@bandoti < Sorry for the late reply.
Thank you for the comments. I'll follow up this in coming 2~3 days.

sparkleholic commented on ggml/src/ggml-vulkan/CMakeLists.txt:
```diff
@@ -1,5 +1,41 @@
+        else()
+            message(WARNING "Neither MSVC nor clang found")
+        endif()
+    elseif(CMAKE_HOST_SYSTEM_NAME STREQUAL "Linux")
```
addressed.

sparkleholic commented on ggml/src/ggml-vulkan/CMakeLists.txt:
```diff
@@ -45,16 +81,57 @@ if (Vulkan_FOUND)
+            else()
+                message(STATUS "Host compiler: ${HOST_C_COMPILER} ${HOST_CXX_COMPILER}")
+            endif()
+            file(WRITE ${CMAKE_BINARY_DIR}/host_toolchain.cmake
```
addressed.

sparkleholic committed:
refactor: ggml: Improve vulkan-shaders-gen toolchain setup

Use configure_file to generate host_toolchain.cmake from template

sparkleholic commented:
Rebased branch to latest master.

bandoti commented:
@sparkleholic No worries on the delay! Your contribution is greatly appreciated :)

It appears as though the test failure is unrelated to this. I'm wondering if merging the latest and rerunning the tests will clear the status. I have only one minor question in the review.

bandoti commented on ggml/src/ggml-vulkan/vulkan-shaders/CMakeLists.txt:
```diff
@@ -3,7 +3,11 @@ find_package(Vulkan COMPONENTS glslc REQUIRED)
 set(TARGET vulkan-shaders-gen)
 add_executable(${TARGET} vulkan-shaders-gen.cpp)
-install(TARGETS ${TARGET} RUNTIME)
+if (${CMAKE_RUNTIME_OUTPUT_DIRECTORY})
```
Is the intention here to be the variable value or the name?
```
if (CMAKE_RUNTIME_OUTPUT_DIRECTORY)
```

sparkleholic replied:
Removed unnecessary code.

bandoti commented:
@sparkleholic I have submitted a pull request to your master_fix branch which updates to latest and resolves a minor merge conflict. Please merge it into your branch so we can get the CI fixed here and merge these changes in!

Here's the PR for reference: https://github.com/sparkleholic/llama.cpp/pull/2

sparkleholic commented:
> @sparkleholic I have submitted a pull request to your master_fix branch which updates to latest and resolves a minor merge conflict. Please merge it into your branch so we can get the CI fixed here and merge these changes in!
> 
> Here's the PR for reference: [sparkleholic#2](https://github.com/sparkleholic/llama.cpp/pull/2)

@bandoti < Thanks a lot!
I'l like to merge your PR into my branch first to proceed this PR.
![image](https://github.com/user-attachments/assets/0cfcb94e-01fd-4605-8998-d93920060215)
Could you confirm which option should be selected for this?

bandoti commented:
@sparkleholic A regular merge commit (the default) should be good. Our changes here will get squashed once they merge it into master. 😊

0cc4m commented:
You seem to have gathered a lot of unrelated changes here, the PR needs to be rebased before it can be merged.

sparkleholic committed:
fix: ggml: Fix compile error

Fix compile error not finding vulkan-shaders-gen

sparkleholic committed:
fix: vulkan-shaders-gen build and path handling

Fix build issues with vulkan-shaders-gen:
- Add target dependency for correct build order
- Use CMAKE_HOST_SYSTEM_NAME for executable suffix
- Fix MSVC output directory in host toolchain
- Normalize path handling for cross-compilation

sparkleholic committed:
Merge pull request #3 from sparkleholic/master_fix2

fix: normalize vulkan-shaders-gen output path for MSVC

bandoti commented on ggml/src/ggml-vulkan/CMakeLists.txt:
```diff
@@ -1,5 +1,39 @@
+function(detect_host_compiler)
+    if (CMAKE_HOST_SYSTEM_NAME STREQUAL "Windows")
+        find_program(MSVC_COMPILER cl)
+        find_program(GNU_C_COMPILER gcc)
```
I think we probably want `NO_CMAKE_FIND_ROOT_PATH` for these `find_program` calls to ensure that when the cross-compile search path is rooted somewhere else, this will ensure that the proper location is searched.

sparkleholic replied:
@bandoti < It's good point.
Yes, `NO_CMAKE_FIND_ROOT_PATH` seems to be required here.
I'll address it.

sparkleholic committed:
fix: improve host compiler detection in vulkan shader build

Improve host compiler detection for vulkan shader generation:
- Add NO_CMAKE_FIND_ROOT_PATH to all compiler searches
- Consolidate compiler detection logic
- Fix Windows-specific MSVC detection
- Ensure correct compiler search in cross-compilation

mtmcp reviewed (changes requested):
Everything seems good to me except my one comment on `detect_host_compiler`. After that (whether the suggestion works or not and we keep it as-is) I am happy to move forward!

sparkleholic commented on ggml/src/ggml-vulkan/CMakeLists.txt:
```diff
@@ -1,5 +1,34 @@
+
 find_package(Vulkan COMPONENTS glslc REQUIRED)
 
+function(detect_host_compiler)
```
Your suggested code looks much more concise and better. However, on Windows where MSVC_COMPILER exists, it seems like it would be better to search for cl with the highest priority. To do this, I modified the compiler listing order of find_program to put cl first instead of searching for MSVC_COMPILER.

sparkleholic replied:
@bandoti <
On Windows, I think the MSVC compiler would be the most stable one to look for, but I'm not sure if it's necessary. Would you please give me your opinion on which of the options below would be better.

* option1
     * Searching order: gcc clang cl in all platforms.
     * In this case, we can remove detect_host_compiler function, could write only two lines in the location calling that function.
```
function(detect_host_compiler)
    find_program(HOST_C_COMPILER NAMES gcc clang cl NO_CMAKE_FIND_ROOT_PATH)
    find_program(HOST_CXX_COMPILER NAMES g++ clang++ cl NO_CMAKE_FIND_ROOT_PATH)
endfunction()
```
* option2
     * Searching order: cl gcc clang in all platforms.
However, there is a downside to assuming an unnecessary process of finding cl even on non-Windows platforms.
     * In this case, we can remove detect_host_compiler function, could write only two lines in the location calling that function.
```
function(detect_host_compiler)
    find_program(HOST_C_COMPILER NAMES cl gcc clang NO_CMAKE_FIND_ROOT_PATH)
    find_program(HOST_CXX_COMPILER NAMES cl g++ clang++ NO_CMAKE_FIND_ROOT_PATH)
endfunction()
```
* option3
For Windows only, the cl compiler is found and applied with the highest priority.
```
function(detect_host_compiler)
    if (CMAKE_HOST_SYSTEM_NAME STREQUAL "Windows")
        find_program(MSVC_COMPILER cl NO_CMAKE_FIND_ROOT_PATH)
        if (MSVC_COMPILER)
            set(HOST_C_COMPILER "${MSVC_COMPILER}" PARENT_SCOPE)
            set(HOST_CXX_COMPILER "${MSVC_COMPILER}" PARENT_SCOPE)
            return()
        endif()
    endif()
    find_program(HOST_C_COMPILER NAMES gcc clang NO_CMAKE_FIND_ROOT_PATH)
    find_program(HOST_CXX_COMPILER NAMES g++ clang++ NO_CMAKE_FIND_ROOT_PATH)
endfunction()
```

sparkleholic committed:
refactor: Simplify CMake function for detecting host compiler

Simplified the CMake function to improve the process of detecting the host compiler.

sparkleholic commented:
Please hold off on merging this patch for a moment. I'm checking my cross-compilation environment one last time.

sparkleholic committed:
fix: Remove unnecessary Vulkan library linkage in CMakeLists.txt

Since `vulkan-shader-gen.cpp` only requires the `glslc` executable
and not the Vulkan headers or libraries, CMakeLists.txt needs to
be corrected.
(See: ecc93d0558fc3ecb8a5af69d2ece02fae4710ade)

sparkleholic commented:
FYI: @jeffbolznv 
Added a necessary change for cross-compiling environment.
https://github.com/ggerganov/llama.cpp/pull/10448/commits/37d0cb6e848ac4c96066f41c3e8ac2e629bd8ba9

sparkleholic commented:
@bandoti <
Ready to merge.

bandoti commented on ggml/src/ggml-vulkan/vulkan-shaders/CMakeLists.txt:
```diff
@@ -1,9 +1,11 @@
 find_package (Threads REQUIRED)
-find_package(Vulkan COMPONENTS glslc REQUIRED)
+find_program(GLSLC_EXECUTABLE glslc)
```
We probably need the `NO_CMAKE_FIND_ROOT_PATH` option here as well.

sparkleholic replied:
I think this line can be used with CMAKE_FIND_ROOT_PATH, depending on the user's intention. For example, the user could set their own CMAKE_TOOLCHAIN_FILE using the GGML_SHADERS_GEN_TOOLCHAIN variable instead of using ggml/src/ggml-vulkan/host_toolchain.cmake.in, and define CMAKE_FIND_ROOT_PATH to specify a path specific to the host system.

bandoti replied:
That's fair. Okay, we can follow up with new PR if this causes an issue. :)

mtmcp reviewed (approved):
Looks good to me!

sparkleholic commented:
@bandoti 
This PR is still opened.

0cc4m commented:
> @bandoti , @ngxson This PR is still open...

It needs approval from a contributor. I'll do a quick test that everything runs as intended on a regular setup later today and merge if it's fine.

ngxson commented:
This change is not related to my part. I was added because you pushed a merge commit or rebase.

Please ask for people who maintains the vulkan backend.

ggerganov commented on ggml/src/ggml-vulkan/host_toolchain.cmake.in:
This file should be moved to a `cmake` subfolder and use dash instead of underscore: `ggml/src/ggml-vulkan/cmake/host-toolchain.cmake.in`

sparkleholic committed:
refactor: Rename host_toolchain.cmake.in

- Rename host_toolchain.cmake.in to cmake/host-toolchain.cmake.in

ggerganov commented on ggml/CMakeLists.txt:
```diff
@@ -185,6 +185,9 @@ option(GGML_OPENCL_PROFILING                "ggml: use OpenCL profiling (increas
 option(GGML_OPENCL_USE_ADRENO_KERNELS       "ggml: use optimized kernels for Adreno"          ON)
 
+# toolchain for vulkan-shaders-gen
+set   (GGML_SHADERS_GEN_TOOLCHAIN "" CACHE FILEPATH "ggml: toolchain file for vulkan-shaders-gen")
```
This option should have a `GGML_VULKAN` prefix. For example, it can be named `GGML_VULKAN_SHADERS_GEN_TOOLCHAIN`

0cc4m replied:
Do you have more comments or is that it? The PR has been open for way too long now and I'd like to get it merged.

ggerganov replied:
No, merge at will.

sparkleholic replied:
Addressed. Thanks.

sparkleholic committed:
refactor: GGML_VULKAN_SHADERS_GEN_TOOLCHAIN

Rename the macro GGML_SHADERS_GEN_TOOLCHAIN to GGML_VULKAN_SHADERS_GEN_TOOLCHAIN

0cc4m reviewed (approved):
Thank you for your patience with all the feedback. I tested it and found no issues.
